src/chainstate/burn/db/burndb.rs

```diff
@@ -604,16 +681,268 @@ impl BurnDB {
+        let arrival_index = BurnDB::get_arrival_index(tx)?;
+        let args : &[&dyn ToSql] = &[&u64_to_sql(stacks_block_height)?, &u64_to_sql(arrival_index + 1)?, burn_header_hash, stacks_block_hash];
+
+        tx.execute("UPDATE snapshots SET stacks_block_accepted = 1, stacks_block_height = ?1, arrival_index = ?2 WHERE burn_header_hash = ?3 AND winning_stacks_block_hash = ?4", args)
+            .map_err(db_error::SqliteError)?;
+
+        let parent_key = BurnDB::make_stacks_block_index_key(parent_stacks_block_hash);
+
+        // update memoized canonical stacks chain tip on the canonical burn chain tip if this block
+        // extends it.
+        if burn_tip.canonical_stacks_tip_hash == *parent_stacks_block_hash {
+            // This block builds off of the memoized canonical stacks chain tip information we
+            // already have.
+            // Memoize this tip to the canonical burn chain snapshot.
+            if stacks_block_height > 0 {
+                assert_eq!(burn_tip.canonical_stacks_tip_height + 1, stacks_block_height);
+            }
+            else {
+                assert_eq!(stacks_block_hash, &FIRST_STACKS_BLOCK_HASH);
+            }
+            debug!("Accepted Stacks block {}/{} builds on the memoized canonical chain tip ({})", burn_header_hash, stacks_block_hash, &burn_tip.burn_header_hash);
+            let args : &[&dyn ToSql] = &[burn_header_hash, stacks_block_hash, &u64_to_sql(stacks_block_height)?, &burn_tip.burn_header_hash];
+            tx.execute("UPDATE snapshots SET canonical_stacks_tip_burn_hash = ?1, canonical_stacks_tip_hash = ?2, canonical_stacks_tip_height = ?3
+                        WHERE burn_header_hash = ?4", args)
+                .map_err(db_error::SqliteError)?;
+        
+            BurnDB::insert_accepted_stacks_block_pointer(tx, &burn_tip.burn_header_hash, burn_header_hash, stacks_block_hash, stacks_block_height)?;
+        }
+        else {
+            // see if this block builds off of a Stacks block mined on this burnchain fork
+            let height_opt = match BurnDB::get_accepted_stacks_block_pointer(tx, &burn_tip.burn_header_hash, parent_stacks_block_hash)? {
+                // this block builds on a block accepted _after_ this burn chain tip was processed?
+                Some(accepted_header) => Some(accepted_header.height),
+                None => match BurnDB::index_value_get(tx, &burn_tip.burn_header_hash, &parent_key)? {
+                    // this block builds on a block accepted _before_ this burn chain tip was processed?
+                    Some(height_str) => Some(height_str.parse::<u64>().expect(&format!("BUG: MARF stacks block key '{}' does not map to a u64", parent_key))),
+                    None => None
+                }
+            };
+            match height_opt {
+                Some(height) => {
+                    if stacks_block_height > height && stacks_block_height > burn_tip.canonical_stacks_tip_height {
```

Comment:
If `stacks_block_height <= height`, that should be some sort of corruption error, right?

Reply:
Yes, I believe so.  If this were true, then we would have processed a Stacks block out-of-order on this burnchain fork (which should have been prevented by a correct implementation of `StacksChainState::process_next_staging_block`).  Will change to a runtime assert and see if it makes a difference.

Reply:
Fixed in 2881929